x/mongo/driver/topology/connection.go

```diff
@@ -467,21 +467,21 @@ func (c *connection) read(ctx context.Context, dst []byte) (bytesRead []byte, er
 		return nil, errResponseTooLarge.Error(), errResponseTooLarge
 	}
 
-	if int(size) > cap(dst) {
+	if dst == nil || int(size) > cap(*dst) {
 		// Since we can't grow this slice without allocating, just allocate an entirely new slice.
-		dst = make([]byte, 0, size)
+		*dst = make([]byte, 0, size)
```

Comment:
We have to couple the resource management logic here because we are using a raw slice (pointer). Instead, I suggest updating our interfaces to use io.Reader/io.Writer alike struct so the resource can be managed independently.